### PR TITLE
tests for hammer job-invocation, fixes #3822

### LINKED ISSUE #3822: add tests for hammer job-template and job-invocation
```
hammer job-template -h
Usage:
    hammer job-template [OPTIONS] SUBCOMMAND [ARG] ...

Subcommands:
 create                        Create a job template
 delete                        Delete a job template
 dump                          View job template content
 info                          Show job template details
 list                          List job templates
 update                        Update a job template
```

```
hammer job-invocation -h
Usage:
    hammer job-invocation [OPTIONS] SUBCOMMAND [ARG] ...

Subcommands:
 create                        Create a job invocation
 info                          Show job invocation
 list                          List job invocations
 output                        View the output for a host
```

UI stubs at #3269

### PR DESCRIPTION
**Pull request by pondrejk**
Test results:

```
nosetests tests/foreman/cli/test_remoteexecution.py:RemoteExecutionTestCase
...
----------------------------------------------------------------------
Ran 3 tests in 118.839s

OK
```

### CODE REVIEW

File: tests/foreman/cli/test_remoteexecution.py [hunk 1]
```diff
@@ -15,19 +15,24 @@
 
 @Upstream: No
 """
-
+from datetime import datetime, timedelta
 from fauxfactory import gen_string
 from robottelo import ssh
+from robottelo.constants import REPOS
```
Comment (oshtaier) [inline tests/foreman/cli/test_remoteexecution.py]: put alphabetically

File: tests/foreman/cli/test_remoteexecution.py [hunk 2]
```diff
@@ -168,7 +173,25 @@ def test_positive_view_dump(self):
...
     """Implements job execution tests in CLI."""
 
-    @stubbed
+    @classmethod
+    def setUpClass(cls):
+        """Create an organization which can be re-used in tests."""
+        super(RemoteExecutionTestCase, cls).setUpClass()
+        cls.organization = make_org()
+        ssh.command(
+            '''echo 'echo "test"' > {0}'''.format(TEMPLATE_FILE)
```
Comment (oshtaier) [inline tests/foreman/cli/test_remoteexecution.py]: In `JobTemplateTestCase` class we used `'''echo '<%= input("command") %>' > {0}'''.format(TEMPLATE_FILE)` as it is standard template for SSH that can be seen through UI
Is it any reason to do that like this?

File: tests/foreman/cli/test_remoteexecution.py [hunk 3]
```diff
@@ -168,7 +173,25 @@ def test_positive_view_dump(self):
...
+        )
+        cls.client = VirtualMachine(distro='rhel71')
```
Comment (oshtaier) [inline tests/foreman/cli/test_remoteexecution.py]: Do not see any removal procedure for that VM. Please, refer to our vm class and context manager:
`with VirtualMachine(distro='rhel71') as client:`
In case you perform that action in `setUpClass` like you do here, please do corresponding destroy action in `tearDownClass`

File: tests/foreman/cli/test_remoteexecution.py [hunk 4]
```diff
@@ -177,8 +200,13 @@ def test_positive_run_default_job_template(self):
 
         @Assert: Verify the job was successfully ran against the host
         """
+        invocation_command = JobInvocation.create({
```
Comment (oshtaier) [inline tests/foreman/cli/test_remoteexecution.py]: Why do we not use factory module for `create()` procedure like we do for all other entities in CLI?

File: tests/foreman/cli/test_remoteexecution.py [hunk 5]
```diff
@@ -197,3 +235,17 @@ def test_positive_run_scheduled_job_template(self):
...
+            system_current_time, '%b %d %Y %I:%M%p')
+        plan_time = (current_time_object + timedelta(seconds=30)).strftime(
+            "%Y-%m-%d %H:%M")
+        invocation_command = JobInvocation.create({
+            'job-template': 'Run Command - SSH Default',
+            'inputs': 'command="ls"',
+            'start-at': plan_time,
+            'search-query': "name ~ {0}".format(self.client.hostname),
+        })
+        sleep(60)
```
Comment (oshtaier) [inline tests/foreman/cli/test_remoteexecution.py]: Static delays is not good approach. Could you find any other idea with a loop for example?

File: tests/foreman/cli/test_remoteexecution.py [hunk 6]
```diff
@@ -197,3 +235,17 @@ def test_positive_run_scheduled_job_template(self):
 
         @Assert: Verify the job was successfully ran after the designated time
         """
+        system_current_time = ssh.command('date +"%b %d %Y %I:%M%p"').stdout[0]
```
Comment (oshtaier) [inline tests/foreman/cli/test_remoteexecution.py]: Any particular reason why we cannot use `datetime.now()` or `datetime.now()+ timedelta(seconds=30)` for all date interactions?

**Comment by oshtaier** [comment]
Also propose to block that PR till https://github.com/SatelliteQE/robottelo/pull/3856 is merged as it is has direct impact on functionality mentioned here.
It doesn't seem that `rhel71` is actual anymore

File: tests/foreman/cli/test_remoteexecution.py (hunk 2 above)
Comment (pondrejk) [inline tests/foreman/cli/test_remoteexecution.py]: Yes, *hammer job-invocation create --file template* requires a template file on disk, so I pre-create it trough ssh, I didn't find a way to go about it differently

Comment (oshtaier) [inline tests/foreman/cli/test_remoteexecution.py]: ok, let me rephrase. Why did you choose template like this?

File: tests/foreman/cli/test_remoteexecution.py (hunk 6 above)
Comment (pondrejk) [inline tests/foreman/cli/test_remoteexecution.py]: That was my first try but I got into a timezone mismatch between the machine from which I ran Robotello and the server with Foreman, so this way I'm getting the server time. On a second thought, maybe it won't be a problem in automation, I'm keen to get educated here :)

File: tests/foreman/cli/test_remoteexecution.py (hunk 2 above)
Comment (pondrejk) [inline tests/foreman/cli/test_remoteexecution.py]: Ah I see, I realized that would be the same as using the predefined template in *test_positive_run_default_job_template*, so I wanted something different for the custom one. I see it's not very sophisticated, I'll look around for someth. better

File: tests/foreman/cli/test_remoteexecution.py [hunk 7]
```diff
@@ -168,7 +174,31 @@ def test_positive_view_dump(self):
...
+        cls.client.register_contenthost(
+            cls.organization['label'],
+            lce='Library'
+        )
+        cls.client.enable_repo(REPOS['rhst7']['id'])
+        cls.client.install_katello_agent()
+        add_remote_execution_ssh_key(cls.client.hostname)
+
+    @classmethod
+    def tearDown(self):
```
Comment (oshtaier) [inline tests/foreman/cli/test_remoteexecution.py]: `tearDownClass?`

File: tests/foreman/cli/test_remoteexecution.py [hunk 8]
```diff
@@ -197,3 +242,22 @@ def test_positive_run_scheduled_job_template(self):
...
+            system_current_time, '%b %d %Y %I:%M%p')
+        plan_time = (current_time_object + timedelta(seconds=30)).strftime(
+            "%Y-%m-%d %H:%M")
+        invocation_command = make_job_invocation({
+            'job-template': 'Run Command - SSH Default',
+            'inputs': 'command="ls"',
+            'start-at': plan_time,
+            'search-query': "name ~ {0}".format(self.client.hostname),
+        })
+        max_attempts = 5
```
Comment (oshtaier) [inline tests/foreman/cli/test_remoteexecution.py]: seems a little bit redundant variable
`for _ in range(5):` probably should suffice our needs

File: tests/foreman/cli/test_remoteexecution.py (hunk 6 above)
Comment (oshtaier) [inline tests/foreman/cli/test_remoteexecution.py]: Let me know your decision here. You can try to use out standalone job on Jenkins for testing if you like to have similar env to our test env.

File: tests/foreman/cli/test_remoteexecution.py [hunk 9]
```diff
@@ -168,7 +174,31 @@ def test_positive_view_dump(self):
...
-    @stubbed
+    @classmethod
+    def setUpClass(cls):
+        """Create an organization which can be re-used in tests."""
+        super(RemoteExecutionTestCase, cls).setUpClass()
+        cls.organization = make_org()
+        ssh.command(
+            '''echo 'getenforce' > {0}'''.format(TEMPLATE_FILE)
+        )
+        cls.client = VirtualMachine(distro=DISTRO_RHEL7)
```
Comment (sghai) [inline tests/foreman/cli/test_remoteexecution.py]: :+1:

File: tests/foreman/cli/test_remoteexecution.py (hunk 6 above)
Comment (pondrejk) [inline tests/foreman/cli/test_remoteexecution.py]: Sorry for delay, if there is no risk in this approach, I'd prefer to keep it as it is, it should work both locally and in automation.

**Comment by sghai** [comment]
@pondrejk: Looks good to me.. Could you please share the tests results again ?

**Comment by sghai** [comment]
@pondrejk : ping ?

**Comment by pondrejk** [comment]
Sorry for the delay:
```
nosetests tests/foreman/cli/test_remoteexecution.py:RemoteExecutionTestCase
...
----------------------------------------------------------------------
Ran 3 tests in 83.060s

OK
```

**Comment by sghai** [comment]
ACK.. looks good. Merging now.
